parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ContextUtil.java

```diff
@@ -251,17 +260,41 @@ public static Configuration getConfiguration(JobContext context) {
     }
   }
 
-  public static Counter getCounter(TaskInputOutputContext context,
-                                   String groupName, String counterName) {
-    return (Counter) invoke(GET_COUNTER_METHOD, context, groupName, counterName);
+  public static Counter getCounter(TaskAttemptContext context, String groupName, String counterName) {
+    Method counterMethod = findCounterMethod(context);
+    return invoke(counterMethod, context, groupName, counterName);
+  }
+
+  public static boolean hasCounterMethod(TaskAttemptContext context) {
+    return findCounterMethod(context) != null;
+  }
+
+  private static Method findCounterMethod(TaskAttemptContext context) {
+    if (context != null) {
+      if (COUNTER_METHODS_BY_CLASS.containsKey(context.getClass())) {
+        return COUNTER_METHODS_BY_CLASS.get(context.getClass());
+      }
+
+      try {
+        Method method = context.getClass().getDeclaredMethod("getCounter", String.class, String.class);
+        if (method.getReturnType().isAssignableFrom(Counter.class)) {
+          COUNTER_METHODS_BY_CLASS.put(context.getClass(), method);
+          return method;
+        }
+      } catch (NoSuchMethodException e) {
+        return null;
+      }
+    }
+
+    return null;
   }
 
   /**
    * Invokes a method and rethrows any exception as runtime exceptions.
    */
-  private static Object invoke(Method method, Object obj, Object... args) {
+  private static <T> T invoke(Method method, Object obj, Object... args) {
     try {
-      return method.invoke(obj, args);
+      return (T)method.invoke(obj, args);
```

Comment:
That would create a warning. The only safe way to do that is casting from the caller.